tests/test_cancellations.py

```diff
@@ -0,0 +1,242 @@
+import typing
+
+import anyio
+import hpack
+import hyperframe
+import pytest
+
+import httpcore
+
+
+class SlowWriteStream(httpcore.AsyncNetworkStream):
+    """
+    A stream that we can use to test cancellations during
+    the request writing.
+    """
+
+    async def write(
+        self, buffer: bytes, timeout: typing.Optional[float] = None
+    ) -> None:
+        await anyio.sleep(999)
+
+    async def aclose(self) -> None:
+        pass
```

Comment:
It seems like good practice to include an await in these noop methods i.e. `await anyio.sleep(0)` to simulate typical yield behavior for the async function.